### PR TITLE
develop/add-journey-record-screen

### PR DESCRIPTION
**Pull request by jeongJ2004**
Corresponding issue: #45

**Please note that** : Since we weren't aware that in 1 PR, it contains 1 task, (code + test), just ONLY for this time, me (Jaeyi), Pablo (Reviewer of the PR and reimplemented the screen) and Lei worked together and tried to merge 2 tasks (implementing JourneyRecord screen -> Jaeyi / implementing tests -> Lei) in one PR.

This PR introduces a series of enhancements to the BrewR app, focusing on the JourneyRecordScreen and the navigation flow. The work ensures that users can view the details of selected journeys, including their uploaded images, while also improving the overall user experience with comprehensive UI tests.

Journey Details Display:
- Feature : The JourneyRecordScreen now observes and displays the selected journey’s details, including the coffee shop name, description, coffee origin, brewing method, taste, rating, and date.

- Feature : Users can now navigate to the JourneyRecordScreen by selecting a journey in the GalleryScreen.

Image Display :
- Feature : The uploaded image for each journey is now displayed within the JourneyRecordScreen instead of a placeholder. The coil library is used for image loading with a smooth crossfade effect.

Date Formatting :
- Feature : The date of each journey is now displayed in a user-friendly dd/MM/yyyy format.

UI Tests for JourneyRecordScreen :

- Feature : Comprehensive UI tests were added to validate the functionality and UI of the JourneyRecordScreen.
Details :
- Test Display of All Components: Verifies that all components (e.g., image, description, origin, etc.) are correctly displayed in the JourneyRecordScreen.
- Test Invalid Navigation : Ensures that no navigation occurs if an invalid journey is selected (e.g., missing details).
- Test Tags : Test tags were added to key components in the screen to facilitate testing.

Navigation Setup for JourneyRecordScreen :
- Feature : The JourneyRecordScreen was added to the app's navigation structure, allowing users to navigate between different parts of the app seamlessly.

RecordJourneyScreen

### CODE REVIEW

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        onSuccess()
+    }.addOnFailureListener {
+        // Log error if the deletion fails
+        it.printStackTrace()
+    }
+}
+
+
+// Function to delete the journey record from Firestore
+fun deleteJourneyFromDatabase(journeyId: String, onSuccess: () -> Unit) {
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:329]: NO!!!!! We have the ViewModel to do that!!!!

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+                )
+            }
+        }
+    )
+}
+
+
+
+// Function to delete the journey and its image
+fun deleteJourney(
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:290]: They can be done in two lines:
deletePicture(journey.imageUrl)
listJourneysViewModel.deleteJourney(journey)

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                    val formattedDate = journey?.date?.toDate()?.let { dateFormat.format(it) } ?: "Unknown Date"
+                    Text(
+                        text = "Date: \n$formattedDate",
+                        style = MaterialTheme.typography.bodyLarge,
+                        fontWeight = FontWeight.Bold,
+                        modifier = Modifier.testTag("date")
+                    )
+                }
+
+                item {
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:241]: Location is no longer a parameter of Journey data class

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun JourneyRecordScreen(
+    listJourneysViewModel: ListJourneysViewModel = viewModel(factory = ListJourneysViewModel.Factory),
+    navigationActions: NavigationActions
+) {
+    val journey = listJourneysViewModel.selectedJourney.collectAsState().value
+    var showDeleteDialog by remember { mutableStateOf(false) } // State to control dialog visibility
+
+    Scaffold(
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:64]: modifier = Modifier.testTag("journeyRecordScreen")

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        topBar = {
+            TopAppBar(
+                title = { Text("Journey Record", fontWeight = FontWeight.Bold) },
+                navigationIcon = {
+                    IconButton(onClick = { navigationActions.goBack() }) {
+                        Icon(
+                            imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
+                            contentDescription = "Back"
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:72]: contentDescription = "goBackButton"

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    navigationActions: NavigationActions
+) {
+    val journey = listJourneysViewModel.selectedJourney.collectAsState().value
+    var showDeleteDialog by remember { mutableStateOf(false) } // State to control dialog visibility
+
+    Scaffold(
+        topBar = {
+            TopAppBar(
+                title = { Text("Journey Record", fontWeight = FontWeight.Bold) },
+                navigationIcon = {
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:68]: test tag?

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+                    IconButton(onClick = { navigationActions.goBack() }) {
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:69]: test tag?

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+                            imageVector = Icons.Filled.Delete,
+                            contentDescription = "Delete Journey",
+                            tint = Color.Red
+                        )
+                    }
+                },
+                modifier = Modifier.testTag("journeyRecordScreen")
+            )
+        },
+        floatingActionButton = {
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:91]: test tag?

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+                    Spacer(modifier = Modifier.height(8.dp))
+
+                    // Image placeholder or uploaded image
+                    Box(
+                        modifier = Modifier
+                            .fillMaxWidth()
+                            .height(200.dp)
+                            .border(1.dp, Color.Gray)
+                    ) {
+                        if (journey?.imageUrl?.isNotEmpty() == true) {
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:145]: if (journey?.imageUrl != null)

File: app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+                            Image(
```
Comment (veridisyndrome) [inline app/src/main/java/com/android/brewr/ui/overview/JourneyRecord.kt:146]: test tag?

**Review by veridisyndrome — CHANGES_REQUESTED** [review]
I will refractor the code according to the comments

**Review by veridisyndrome — COMMENTED** [review]
LGTM

**Comment by LeiChen5** [comment]
All checks have passed, the line coverage on new code is 96%.

**Review by veridisyndrome — APPROVED** [review]
LGTM!!

**Pull request merged**